apm-collector/apm-collector-baseline/collector-baseline-computing-define/src/main/java/org/skywalking/apm/collector/baseline/computing/service/ComputingService.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2017, OpenSkywalking Organization All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Project repository: https://github.com/OpenSkywalking/skywalking
+ */
+
+package org.skywalking.apm.collector.baseline.computing.service;
+
+import org.skywalking.apm.collector.baseline.computing.Configuration;
+import org.skywalking.apm.collector.baseline.computing.Metric;
+import org.skywalking.apm.collector.core.module.Service;
+
+import java.util.List;
+
+/**
+ * @author zhang-chen
+ */
+public interface ComputingService extends Service {
+
+    List<Metric> compute(List<Metric>[] metrics, Configuration conf);
```

Comment:
Every service is based on its **provider**, and `provider` starts up based on its on `prepare`, `start` and `notifyAfterCompleted` stages. Check `ComputingProvider.java` methods. All the configurations come from the yml style config file. So the service consumer will not be care of `Configuration`.So,  I recommend to remove the `Configuration` parameter.

Reply:
Correct me if I misunderstand something. Do you provide two computing ways?
1. computing gaussian distribution
1. computing average

If so, how to make the decision which way will the computing implementation use? I prefer to config this in yml file, too.

Reply:
Let me explain how baseline line computing first.
  There are two steps:
  First step, Baseline service will receive more than three days original performance data as a method parameter. If original data is collected every minute, there are 1440 point on one day data, means there are 1440 Metric Class instance object in List<Metric> parameter, baseline computing program will recompute every daily performance data by Gauss algorithm, in order to make it more smooth.  
  Second step, baseline computing program will calculate average of the same time each day after discard a few (specified by discard parameter) highest/lowest values.
  So actually there will be one provider at present. Configuration parameter can be removed instead of provide default value in ComputingServiceImpl. 
  Configuration parameter is really a problem, if remain it, it is hardly provide another implementation. If remove it, service consumer can not control specific detail on baseline computing.

Reply:
I see what you mean, set configuration parameter into yml style config file, and provider will receive it on prepare phase, that is great idea.

Reply:
Thanks for you explaination. I have a further question, I prefer the Metric Class can be more explicit than now. May be you show give a demonstration about how does implementation understand(group) the input Metric list.

e.g.  For `If original data is collected every minute, there are 1440 point on one day data, means there are 1440 Metric Class instance object in List parameter`
1. What is the output after computing? 
1. If the service user set the list with 1440 points, how to you know if based on every minute of a day? For now, I think you don't, then how to make sure your group strategy is right. 1440 can be consider as 60 minutes * 24 hours, or 120 minutes(2hours) * 12 bucket. 

 That will make user easier to understand the service means.

Configuration should not be concern by service user, it depends on service provider. That is the principle for module isolation. Because more by more algorithms can be shared.

Reply:
1. The output is a List<Metric> which has same size as input List 
  2. As you said, I do not know and care how many Metric class instance in List, just demand every List in Array has same size. At the beginning, I thought about set a "timestamp" filed into Metric class, so I can find out the data on the same time in different daily data(different daily data refer to different List<Metric> in Array). Then I think I am not supposed to know that, just assume every same index object in different List<Metric> have the same time in a day.  That is why I still consider which is better between Metric[][] and List<Metric>[] as the parameter, do you have some suggestion for the parameter type?

Reply:
From my experience, I will recommend no matter Metric[][] or List[], that is too abstract. You should build a concept/class, maybe named `RawMetric`, for Metric array list, then set a list of that `RawMetric`. And then the returning will be `Baseline` class. By that, it makes things more clear. What do you think?

Reply:
you mean like List<RawMetric> as parameter, RawMetric can be daily data? I wanna service consumer clearly know every RawMetric in List must have same length, if not, it will occur unexpected exception in computing baseline. That is why I consider to use Metric[][] as parameter to make sure that service consumer provide correct data format.  
  If we define input parameter as List<RawMetric>, output could also be RawMetric, because baseline and original data have a same data struct. Only different is baseline looks more smooth in chart because it is average of history data.

Reply:
I understand a concern, but by using `RawMetric`, you can use a `RawMetricBuilder(length)`, with provide `RawMetricBuilder#receive` to make sure the length. After all, this is only a design pattern, right, it is not important.

I think @peng-yongsheng will drop a more high level discussion at the service. Let's wait a moment.

Reply:
look forward to some suggestion from @peng-yongsheng , in the mean time, I will keep thinking about how to make it easier to understand